orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -0,0 +1,501 @@
+                "module checkpoint state has null train_session_data_transfer_mgr.");
+    ORT_RETURN_IF_ERROR(SaveRuntimeTensors(
+        GetTensorProtoFilePath(parameter_folder_path, k_param_root_prefix),
+        GetExternalDataFilePath(parameter_folder_path, k_param_root_prefix),
+        *module_states.train_session_data_transfer_mgr_,
+        model_parameter_ort_values));
+  }
+
+  return Status::OK();
+}
+
+Status CheckpointUtils::OrtSaveOptimizerStatesInternal(OptimizerCheckpointStates& optimizer_states,
+                                                       const PathString& checkpoint_path) {
+  if (optimizer_states.group_named_optimizer_states.empty()) {
+    return Status::OK();
+  }
+
+  // Write optimizer state tensors files.
+  for (auto& group_named_optimizer_state : optimizer_states.group_named_optimizer_states) {
+    const std::string& group_name = group_named_optimizer_state.first;
+    const std::shared_ptr<GroupOptimizerState>& group_optimizer_state_ptr = group_named_optimizer_state.second;
+    const std::string& cur_group_filename_prefix = StringConcat(k_optimizer_root_prefix, group_name);
+
+    /**
+     * Re-organize optimizer_state_ort_values mapping
+     * Firstly indexed by moment state names; Secondly indexed by parameter names.
+     */
+    std::unordered_map<std::string, std::unordered_map<std::string, OrtValue>> optimizer_state_ort_values;
+    for (const std::pair<std::string, ParameterOptimizerState>&
+             param_named_optimizer_state : group_optimizer_state_ptr->param_named_optimizer_states_) {
+      const std::string& param_name = param_named_optimizer_state.first;
+      const auto& param_optimizer_state = param_named_optimizer_state.second;
+
+      for (const std::pair<std::string, std::shared_ptr<OrtValue>>&
+               m_state : param_optimizer_state.states_) {
+        const std::string& m_state_name = m_state.first;
+        const std::shared_ptr<OrtValue>& m_state_val = m_state.second;
+
+        if (optimizer_state_ort_values.find(m_state_name) == optimizer_state_ort_values.end()) {
+          std::unordered_map<std::string, OrtValue> param_name_to_ortvalue{{param_name, *(m_state_val)}};
+          optimizer_state_ort_values.insert({m_state_name, param_name_to_ortvalue});
+        } else {
+          optimizer_state_ort_values[m_state_name].insert({param_name, *(m_state_val)});
+        }
+      }
+    }
+
+    /**
+     * Save each optimizer state (of all parameters) into single file.
+     * For example: save "momentum_1" of all parameters into one file.
+     */
+    for (auto& pair : optimizer_state_ort_values) {
+      const auto& state_name = pair.first;
+      const std::unordered_map<std::string, OrtValue>& param_name_to_ortvalue = pair.second;
+      const std::string& cur_state_filename_prefix = StringConcat(cur_group_filename_prefix, state_name);
+
+      ORT_ENFORCE(optimizer_states.optimizer_session_data_transfer_mgr_,
```

Comment:
can move this check at the beginning (before doing all the work reordering optim state) and out of the loop

Reply:
fixed.